File: topk-rs/tests/test_datasets.rs
use test_context::test_context;
use topk_rs::Error;

mod utils;
use utils::ProjectTestContext;

#[test_context(ProjectTestContext)]
#[tokio::test]
async fn test_list_datasets(ctx: &mut ProjectTestContext) {
    let create = ctx
        .client
        .datasets()
        .create(ctx.wrap("test"))
        .await
        .expect("could not create dataset");

    let list = ctx
        .client
        .datasets()
        .list()
        .await
        .expect("could not list datasets");

    assert!(list
        .datasets
        .iter()
        .any(|d| d.name == create.dataset().unwrap().name));
}

#[test_context(ProjectTestContext)]
#[tokio::test]
async fn test_create_dataset(ctx: &mut ProjectTestContext) {
    let create = ctx
        .client
        .datasets()
        .create(ctx.wrap("test"))
        .await
        .expect("could not create dataset");

    let list = ctx
        .client
        .datasets()
        .list()
        .await
        .expect("could not list datasets");

    assert!(list
        .datasets
        .iter()
        .any(|d| d.name == create.dataset().unwrap().name));
}

#[test_context(ProjectTestContext)]
#[tokio::test]
async fn test_create_duplicate_dataset(ctx: &mut ProjectTestContext) {
    ctx.client
        .datasets()
        .create(ctx.wrap("test"))
        .await
        .expect("could not create dataset");

    let err = ctx
        .client
        .datasets()
        .create(ctx.wrap("test"))
        .await
        .expect_err("should not be able to create duplicate dataset");

    assert!(matches!(err, Error::DatasetAlreadyExists));
}

#[test_context(ProjectTestContext)]
#[tokio::test]
async fn test_delete_non_existent_dataset(ctx: &mut ProjectTestContext) {
    let err = ctx
        .client
        .datasets()
        .delete(ctx.wrap("test"))
        .await
        .expect_err("should not be able to delete non-existent dataset");

    assert!(matches!(err, Error::DatasetNotFound));
}

#[test_context(ProjectTestContext)]
#[tokio::test]
async fn test_delete_dataset(ctx: &mut ProjectTestContext) {
    let create = ctx
        .client
        .datasets()
        .create(ctx.wrap("test"))
        .await
        .expect("could not create dataset");

    ctx.client
        .datasets()
        .delete(ctx.wrap("test"))
        .await
        .expect("could not delete dataset");

    let list = ctx
        .client
        .datasets()
        .list()
        .await
        .expect("could not list datasets");

    assert!(!list
        .datasets
        .iter()
        .any(|d| d.name == create.dataset().unwrap().name));
}

#[test_context(ProjectTestContext)]
#[tokio::test]
async fn test_get_dataset(ctx: &mut ProjectTestContext) {
    let err = ctx
        .client
        .datasets()
        .get(ctx.wrap("test"))
        .await
        .expect_err("should not be able to get non-existent dataset");

    assert!(matches!(err, Error::DatasetNotFound));

    let create = ctx
        .client
        .datasets()
        .create(ctx.wrap("test"))
        .await
        .expect("could not create dataset");

    let get = ctx
        .client
        .datasets()
        .get(ctx.wrap("test"))
        .await
        .expect("could not get dataset");

    assert_eq!(get.dataset().unwrap().name, create.dataset().unwrap().name);
}

File: topk-py/tests/test_datasets.py
import pytest
from topk_sdk import error

from . import ProjectContext


def test_list_datasets(ctx: ProjectContext):
    create_resp = ctx.client.datasets().create(ctx.scope("test"))
    print(create_resp)
    d = create_resp.dataset
    response = ctx.client.datasets().list()
    print(response)
    assert d in response.datasets


def test_create_dataset(ctx: ProjectContext):
    create_resp = ctx.client.datasets().create(ctx.scope("test"))
    print(create_resp)
    d = create_resp.dataset
    response = ctx.client.datasets().list()
    print(response)
    assert d in response.datasets


def test_create_duplicate_dataset(ctx: ProjectContext):
    response = ctx.client.datasets().create(ctx.scope("test"))
    print(response)

    with pytest.raises(error.DatasetAlreadyExistsError):
        ctx.client.datasets().create(ctx.scope("test"))


def test_delete_non_existent_dataset(ctx: ProjectContext):
    with pytest.raises(error.DatasetNotFoundError):
        ctx.client.datasets().delete(ctx.scope("test"))


def test_delete_dataset(ctx: ProjectContext):
    d = ctx.client.datasets().create(ctx.scope("test")).dataset
    response = ctx.client.datasets().delete(ctx.scope("test"))
    print(response)

    response = ctx.client.datasets().list()
    print(response)
    assert d not in response.datasets


def test_get_dataset(ctx: ProjectContext):
    # Test getting non-existent dataset
    with pytest.raises(error.DatasetNotFoundError):
        ctx.client.datasets().get(ctx.scope("test"))

    # Create dataset
    create_resp = ctx.client.datasets().create(ctx.scope("test"))
    print(create_resp)
    d = create_resp.dataset

    # Get dataset
    response = ctx.client.datasets().get(ctx.scope("test"))
    print(response)
    assert response.dataset == d
